protobuf/lib/src/protobuf/coded_buffer_reader.dart

```diff
@@ -13,17 +13,30 @@ class CodedBufferReader {
   static const int DEFAULT_SIZE_LIMIT = 64 << 20;
 
   final Uint8List _buffer;
+
+  /// [ByteData] of [_buffer], created once to be able to decode fixed-size
+  /// integers and floats without having to allocate a [ByteData] every time.
+  final ByteData _byteData;
+
   int _bufferPos = 0;
   int _currentLimit = -1;
   int _lastTag = 0;
   int _recursionDepth = 0;
   final int _recursionLimit;
   final int _sizeLimit;
 
-  CodedBufferReader(List<int> buffer,
-      {int recursionLimit = DEFAULT_RECURSION_LIMIT,
-      int sizeLimit = DEFAULT_SIZE_LIMIT})
-      : _buffer = buffer is Uint8List ? buffer : Uint8List.fromList(buffer),
+  factory CodedBufferReader(List<int> buffer,
+          {int recursionLimit = DEFAULT_RECURSION_LIMIT,
+          int sizeLimit = DEFAULT_SIZE_LIMIT}) =>
+      CodedBufferReader._(
+        buffer is Uint8List ? buffer : Uint8List.fromList(buffer),
+        recursionLimit,
+        sizeLimit,
+      );
+
+  CodedBufferReader._(Uint8List buffer, int recursionLimit, int sizeLimit)
+      : _buffer = buffer,
+        _byteData = ByteData.sublistView(buffer),
```

Comment:
With this we're paying for a `ByteData` allocation even if we don't need it when decoding a proto. There may be many protos that don't have float/double/fixed-ints in them.

Consider doing that lazily:
```
late final ByteData _byteData = ByteData.sublistView(_buffer);
```

(then you may also undo the changes to the constructor)

Reply:
Right, but it's one allocation per top-level `fromBuffer` call, nested messages are decoded using the same `CodedBufferReader`.

I'm not sure how expensive the allocation is, but it will be optimizing small messages at the cost of large ones with float etc. fields.

Do you think it's the right trade-off?

Reply:
I was more thinking about dart2js. We know that JS typed data allocations are expensive. Maybe typed data view (like `ByteData`) allocations are not? In that case, feel free to leave like it is.

Reply:
I implemented your suggestion. It didn't make significant different in my benchmarks.